receiver/hostmetricsreceiver/hostmetrics_receiver.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hostmetricsreceiver
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/component/componenterror"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/collector/cpu"
+	hmcomponent "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/component"
+)
+
+// Receiver is the type used to handle metrics from VM metrics.
+type Receiver struct {
+	consumer consumer.MetricsConsumerOld
+
+	startTime time.Time
+	tickerFn  getTickerC
+
+	config     *Config
+	collectors []hmcomponent.Collector
+
+	stopOnce  sync.Once
+	startOnce sync.Once
+
+	done chan struct{}
+}
+
+type getTickerC func() <-chan time.Time
+
+// NewHostMetricsReceiver creates a new set of VM and Process Metrics
+func NewHostMetricsReceiver(
+	logger *zap.Logger,
+	config *Config,
+	consumer consumer.MetricsConsumerOld,
+	tickerFn getTickerC,
+) (*Receiver, error) {
+
+	configs := []struct {
+		config  hmcomponent.CollectorConfig
+		factory hmcomponent.CollectorFactory
+	}{
+		{config: config.CPUConfig, factory: &cpu.Factory{}},
+	}
+
+	collectors := make([]hmcomponent.Collector, 0)
+	for _, cfg := range configs {
+		if reflect.ValueOf(cfg.config).IsNil() {
```

Comment:
This is the code that ignores a metric type if it's not configured. Not 100% about the method for nil checking here (though reflection should be okay since this is only run on initialization)

Reply:
Not sure what are you trying to do here.
You've just defined `configs` inline above, with exactly 1 element. Why do you need to iterate over it?
Do you plan to add other types of configs there later?

P.S. the nil check via reflection is probably due to https://stackoverflow.com/questions/19761393/why-does-go-have-typed-nil

Reply:
Actually this seems like a bug, reflect will fail, see https://play.golang.org/p/DwGJQYip_ZL
So you need to replace it with
```go
if cfg.config != nil {
  ...
}
```

Reply:
1. Yea there are a bunch of metrics to collect so the idea is to add config for other "collectors" as per the PR description. I should probably express this as a list directly in the config so its clearer/tidier

i.e. change the definition of Config from:
```
type Config struct {
	configmodels.ReceiverSettings `mapstructure:",squash"`

	ScrapeInterval time.Duration `mapstructure:"scrape_interval"`
	CPUConfig      *cpu.Config   `mapstructure:"cpu"`
}
```
to something like :
```
type Config struct {
	configmodels.ReceiverSettings `mapstructure:",squash"`

	ScrapeInterval time.Duration `mapstructure:"scrape_interval"`
	CollectorConfigs      []*CollectorConfig   `mapstructure:"collectors"`
}
```

2. I think it's because of Go's distinction between the interface's concrete type being nil and value being nil:
https://medium.com/@mangatmodi/go-check-nil-interface-the-right-way-d142776edef1

Here's a small update to your playground to demonstrate how this works:
https://play.golang.org/p/BkWCOImyqpa

I guess it's potentially a sign of code smell that I need to perform this check, although it seems like a fairly common thing for people to run into. If I change to a list as per #1 this will go away anyway